### PR TITLE
Fix nnq.dropout in vision mobilenetv3 pretrain model

### PR DESCRIPTION
**Pull request by terrychenism**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71438

Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

### CODE REVIEW

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

File: torch/testing/_internal/common_quantization.py [hunk 1]
```diff
@@ -1633,6 +1635,23 @@ def forward(self, x):
...
+        self.fc1 = torch.nn.Linear(5, 1).to(dtype=torch.float)
+        self.fc2 = torch.nn.Linear(1, 10).to(dtype=torch.float)
+
+    def forward(self, x):
+        x = self.quant(x)
+        x = self.fc1(x)
+        x = self.fc2(x)
+        return self.dequant(x)
+
+class ManualDropoutQATModel(torch.nn.Module):
```
Comment (jerryzh168) [inline torch/testing/_internal/common_quantization.py:1654]: we can remove the two linears in the this model I think

File: torch/ao/quantization/fx/quantization_patterns.py [hunk 2]
```diff
@@ -1499,7 +1499,8 @@ def convert(self,
 @register_quant_pattern(torch.nn.AvgPool1d)
 @register_quant_pattern(torch.nn.AvgPool2d)
 @register_quant_pattern(torch.nn.AvgPool3d)
-@register_quant_pattern(torch.nn.Dropout)
+# TODO fix it when codebase is stable
+# @register_quant_pattern(torch.nn.Dropout)
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/quantization_patterns.py:1503]: we can just remove this I think

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

File: torch/testing/_internal/common_quantization.py (hunk 1 above)
Comment (terrychenism) [reply torch/testing/_internal/common_quantization.py:1654]: remove linear op create an empty parameter list, optimizer would fail.

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

File: torch/testing/_internal/common_quantization.py (hunk 1 above)
Comment (jerryzh168) [reply torch/testing/_internal/common_quantization.py:1654]: optional nit: I see, can we move one of the linears then?

File: torch/ao/quantization/fx/quantization_patterns.py [hunk 3]
```diff
@@ -1509,7 +1508,6 @@ def convert(self,
 @register_quant_pattern(torch.adaptive_avg_pool1d)
 @register_quant_pattern(torch.nn.functional.adaptive_avg_pool2d)
 @register_quant_pattern(torch.nn.functional.adaptive_avg_pool3d)
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/quantization_patterns.py:1510]: should we support functional dropout? removing it is breaking the BC as well

File: torch/testing/_internal/common_quantization.py (hunk 1 above)
Comment (terrychenism) [reply torch/testing/_internal/common_quantization.py:1654]: fc2 removed

File: aten/src/ATen/native/quantized/cpu/qdropout.cpp [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+#include <torch/library.h>
+#include <ATen/quantized/Quantizer.h>
+#include <ATen/native/quantized/cpu/quantized_ops.h>
+
+namespace at {
+namespace native {
+
+DEFINE_DISPATCH(qdropout_stub);
+
+Tensor quantized_dropout(
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cpu/qdropout.cpp:12]: I think this op should take a `training` argument, like https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1279

Comment (jerryzh168) [reply aten/src/ATen/native/quantized/cpu/qdropout.cpp:12]: nit: output_scale and output_zero_point should be the last two arguments of this op, otherwise looks good

Comment (jerryzh168) [reply aten/src/ATen/native/quantized/cpu/qdropout.cpp:12]: for the other parameters it would be best to match the order of original dropout op as well

Comment (jerryzh168) [reply aten/src/ATen/native/quantized/cpu/qdropout.cpp:12]: actually what is scale and input_scale?

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
there is some problem with the quantized_dropout signature I think

File: torch/nn/quantized/functional.py [hunk 5]
```diff
@@ -493,6 +493,19 @@ def elu(input: Tensor, scale: float, zero_point: int, alpha: float = 1.) -> Tens
         raise ValueError("Input to 'quantized.elu' must be quantized!")
     return torch.ops.quantized.elu(input, scale, zero_point, alpha)
 
+def dropout(input: Tensor, scale: float, zero_point: int, p: float = 0.5, training: bool = False) -> Tensor:
```
Comment (jerryzh168) [inline torch/nn/quantized/functional.py:496]: this is probably not used, I think we can remove

**Review by jerryzh168 — APPROVED** [review]
looks great, thanks!

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

**Comment by datumbox** [comment]
Thanks for the fix. I hope we can merge soon to fix our broken CI.
